BLD: NumPy 2 compat for wheel builds

tylerjereddy opened the pull request:
* Related to #4619

* We should build against NumPy `2.0.0` so that we're backwards compatible to NumPy `1.x` while still supporting 2.x series. This is simpler to maintain as well. In theory, should support as far back as NumPy `1.19` if we needed it.


PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4620.org.readthedocs.build/en/4620/

<!-- readthedocs-preview mdanalysis end -->

tylerjereddy committed:
BLD: NumPy 2 compat for wheel builds

* Fixes #4619

* We should build against NumPy `2.0.0` so that we're backwards
compatible to NumPy `1.x` while still supporting 2.x series. This
is simpler to maintain as well. In theory, should support
as far back as NumPy `1.19` if we needed it.

tylerjereddy commented:
some of our Windows deps may also not have NumPy 2 compatible releases yet, need to check on that

IAlibay commented on package/pyproject.toml:
```diff
@@ -3,15 +3,13 @@
-    # NumPy version. In that case we just let it be a bare NumPy install
-    "numpy<2.0; python_version>='3.13'",
+    # numpy requirement for wheel builds for distribution on PyPI - building
+    # against 2.x yields wheels that are also compatible with numpy 1.x at
```
Is the range of compatible 1.x numpy versions with 2.0 published somewhere? I.e. checking if we need to bump the minimum NumPy version anywhere.

tylerjereddy replied:
This document isn't too bad: https://numpy.org/doc/stable/dev/depending_on_numpy.html#for-downstream-package-authors. I think we may be "ok" on that front.

You can deduce the stability back to `1.25.x` and then to `1.19.x` based on that document. It looks like we have Python `3.10` and NumPy `1.23.2` as our runtime lower bounds, which is pretty close to where SciPy is (`3.10` and `1.23.5` respectively) at the moment.

tylerjereddy replied:
Let me do a quick check of the current CI failures to see if I can make a concrete suggestion for those as well.

IAlibay reviewed (commented):
Thanks @tylerjereddy - just the one question.

I have a suspicion some of our gh actions CI won't be picking this at all since we do a lot of isolation disabling here and there, but if the azure pipelines CI works then it should be good.

tylerjereddy commented:
For `pytng` errors, I think the cross-listed PR is roughly what we'd want to do there. Even if there's no bandwidth for a release for `pytng` (let alone MDA) for a bit, we could perhaps point to the GitHub `master` branch of `pytng` for installs in our CI here (i.e., point `pip` at the GitHub repo) to keep that testing active rather than disabling `pytng` testing here.

There may be a few other test failures here to check, but the `pytng` ones stood out first.

I've adjusted the original comment to not close the matching issue, since this may be a step in the right direction, but we may need to check other deps/things a bit still.

tylerjereddy commented:
I restarted the failed Azure jobs to see if the new `pytng` release helps the situation a bit.

tylerjereddy commented:
Looking at the logs, in terms of our dependencies, I'd say `parmed` seems to have moderate issues with NumPy 2 (they need to adjust their array copy semantics a bit like we did). `RDKit` seems to have a more serious problem, resulting in hard crashes against NumPy 2. Both projects probably need a new release to support.

IAlibay commented:
@tylerjereddy sorry I'm going to word this terribly - these are optional dependencies, is there a path you can see here where we stick to 1.x if you need these optional dependencies and 2+ if you don't? I guess in we'd effectively need two packages with different pins?

My view is that this doesn't seem like it would work, but I've put all of five minutes of thought into it.

tylerjereddy committed:
MAINT, TST: PR 4620 revisions

* Add NumPy 2-related shims aimed at avoiding
control flow that hits `RDKit` or `parmed`, until
they have releases supporting NumPy 2. This mostly
means skipping a bunch of tests that use those
packages for now.

pep8speaks commented:
Hello @tylerjereddy! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:















* In the file [`testsuite/MDAnalysisTests/util.py`](https://github.com/MDAnalysis/mdanalysis/blob/7fd4925a4ac9a50c710bcbc5c046275c18684272/testsuite/MDAnalysisTests/util.py):

> [Line 42:1](https://github.com/MDAnalysis/mdanalysis/blob/7fd4925a4ac9a50c710bcbc5c046275c18684272/testsuite/MDAnalysisTests/util.py#L42): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 43:1](https://github.com/MDAnalysis/mdanalysis/blob/7fd4925a4ac9a50c710bcbc5c046275c18684272/testsuite/MDAnalysisTests/util.py#L43): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file



##### Comment last updated at 2024-06-22 13:59:40 UTC

tylerjereddy commented:
@IAlibay I think the best I can suggest is to skip the related tests/imports for `parmed` and `rdkit` under NumPy 2 until they have compatible releases available.

I pushed in a sample commit that seems to allow the testsuite to pass locally with NumPy 2. Looking at the results in CI here, the only test failure is `test_creating_multiple_universe_without_offset`, which I believe is a known issue/flake.

That doesn't mean the MDA team will like this, but I pushed it in so you can take a look. Maybe not too painful?

IAlibay commented:
> @IAlibay I think the best I can suggest is to skip the related tests/imports for `parmed` and `rdkit` under NumPy 2 until they have compatible releases available.
> 
> I pushed in a sample commit that seems to allow the testsuite to pass locally with NumPy 2. Looking at the results in CI here, the only test failure is `test_creating_multiple_universe_without_offset`, which I believe is a known issue/flake.
> 
> That doesn't mean the MDA team will like this, but I pushed it in so you can take a look. Maybe not too painful?

O wow thanks a bunch for digging into this @tylerjereddy !
I don't mind this solution, it's not "clean", but it's the "cleanest" I think we can do here - rdkit will update eventually, parmed might not and there's nothing we can do about that.

IAlibay committed:
Update ParmEd.py

IAlibay commented on package/MDAnalysis/converters/ParmEd.py:
```diff
@@ -168,11 +170,20 @@ def convert(self, obj):
-            import parmed as pmd
+            # TODO: remove this guard when parmed has a release
+            # that supports NumPy 2
+            if NumpyVersion(np.__version__) < "2.0.0":
```
I just added this extra guard here since we were doing the same for RDKit.

IAlibay committed:
Update CHANGELOG

IAlibay reviewed (approved):
Thanks so much @tylerjereddy ! I hope you don't mind I took the liberty of pushing a few commits directly to make sure we can get this merged in asap.

IAlibay committed:
Update ParmEd.py

IAlibay commented:
Thanks @tylerjereddy - going with the merge!
